### PR TITLE
Reduce the number of requests to solana rpc

### PR DESCRIPTION
**Pull request by anton-lisanin**
- Request multiple accounts at the same time
- Confirm multiple transactions at the same time
- Send multiple transactions at the same time, when writing to holder account

### CODE REVIEW

**Review by gigimon — DISMISSED** [review]
looks good

File: proxy/common_neon/solana_interactor.py [hunk 1]
```diff
@@ -109,12 +139,34 @@ def send_transaction_unconfirmed(self, txn: Transaction):
...
+
+        for transaction in transactions:
+            transaction.recent_blockhash = blockhash
+            transaction.sign(self.signer)
+        
+            base64_transaction = base64.b64encode(transaction.serialize()).decode("utf-8")
+            params = (base64_transaction, {"skipPreflight": True, "encoding": "base64"})
+            request = self.client._provider.json_encode({"jsonrpc": "2.0", "id": "1", "method": "sendTransaction", "params": params})
+
+            connection = HTTPSConnection(url.hostname, url.port) if url.scheme == "https" else HTTPConnection(url.hostname, url.port)
```
Comment (s-medvedev) [inline proxy/common_neon/solana_interactor.py]: Initializing of SSL connection takes some amount of time. Have you tried to use batch JSON RPC request to Solana https://www.jsonrpc.org/specification#batch?

File: proxy/common_neon/solana_interactor.py [hunk 2]
```diff
@@ -152,13 +204,46 @@ def confirm_transaction(self, tx_sig, confirmations=0):
...
+
+        raise RuntimeError("could not confirm transactions: ", signatures)
 
     def collect_results(self, receipts, eth_trx=None, reason=None):
+        self.confirm_multiple_transactions(receipts)
+
         results = []
         for rcpt in receipts:
-            results.append(self.collect_result(rcpt, eth_trx, reason))
+            result = self.client.get_confirmed_transaction(rcpt)
```
Comment (s-medvedev) [inline proxy/common_neon/solana_interactor.py]: You can use JSON RPC batch mode here. I tried it on the devnet Solana node:
```
curl -s --header "Content-Type: application/json" --data '[{"method":"getConfirmedTransaction","params":["4yiHB5qMzdamVuKprGo8yFd6w5WT8W2LEapz33GEyWYFo87VBayiWuFiL5z4H5LBJw9XpcdMaWsTGLaX4FPmfzCQ","base64"],"id":93,"jsonrpc":"2.0"},{"method":"getConfirmedTransaction","params":["MzJc9ELCk2Jx47pqRbUzhwHUN54GXeBwsR4cgiL21wpVbFWaRYUUHWeL4QdH6inUkzb6cb1JgtfDoM9WmjvYF1o","base64"],"id":94,"jsonrpc":"2.0"}]' https://api.devnet.solana.com/
```

File: proxy/common_neon/solana_interactor.py (hunk 1 above)
Comment (otselnik) [inline proxy/common_neon/solana_interactor.py]: It is doubtful that sequentially establishing multiple connections will be faster than sending multiple requests with `skipPreflight`.

File: proxy/common_neon/transaction_sender.py [hunk 3]
```diff
@@ -342,32 +339,26 @@ def __init__(self, solana_interactor: SolanaInteractor, neon_instruction: NeonIn
...
-        length = len(self.create_acc_trx.instructions)
-        if length == 0:
-            return
-        logger.debug(f"Create account for trx: {length}")
-        precall_txs = Transaction()
-        precall_txs.add(self.create_acc_trx)
-        result = self.sender.send_measured_transaction(precall_txs, self.eth_trx, 'CreateAccountsForTrx')
-        if check_for_errors(result):
-            raise Exception("Failed to create account for trx")
-        self.create_acc_trx = Transaction()
```
Comment (otselnik) [inline proxy/common_neon/transaction_sender.py:367]: This line is needed to be returned in `call_signed_iterative_combined`.
See https://github.com/neonlabsorg/proxy-model.py/issues/364

File: proxy/common_neon/transaction_sender.py [hunk 4]
```diff
@@ -342,32 +339,26 @@ def __init__(self, solana_interactor: SolanaInteractor, neon_instruction: NeonIn
...
 
     def call_signed_with_holder_combined(self):
-        self.write_trx_to_holder_account()
-        self.create_accounts_for_trx()
+        precall_transactions = self.make_write_to_holder_account_trx()
+        if len(self.create_acc_trx.instructions) > 0:
+            precall_transactions.append(self.create_acc_trx)
+        signatures = self.sender.send_multiple_transactions_unconfirmed(precall_transactions)
+        self.sender.confirm_multiple_transactions(signatures)
+
```
Comment (otselnik) [inline proxy/common_neon/transaction_sender.py]: It is a common situation in Solana to lose some transactions in sequence. Here the retry logic is lost. 
Wouldn't it be better to pass `create_acc_trx` to the `write_to_holder_account_trx` function and work with it in the same way as with write instructions.

File: proxy/common_neon/solana_interactor.py [hunk 5]
```diff
@@ -152,13 +204,46 @@ def confirm_transaction(self, tx_sig, confirmations=0):
...
+            update_transaction_cost(result, eth_trx, reason)
+            results.append(result)
+
         return results
 
+    def collect_result(self, reciept, eth_trx, reason=None):
+        self.confirm_transaction(reciept)
```
Comment (otselnik) [inline proxy/common_neon/solana_interactor.py]: `confirm_transaction` a special case of writing `confirm_multiple_transactions`.
We can get rid of `confirm_transaction` and use `self.confirm_multiple_transactions([reciept])` instead.

**Review by otselnik — COMMENTED** [review]
Great, but looks unfinished without optimizing [call_continue_bucked](https://github.com/neonlabsorg/proxy-model.py/blob/combine_solana_rcp_requests/proxy/common_neon/transaction_sender.py#L436) in a way like write to holder account.

File: proxy/common_neon/transaction_sender.py (hunk 3 above)
Comment (anton-lisanin) [reply proxy/common_neon/transaction_sender.py:367]: Returned

File: proxy/common_neon/solana_interactor.py (hunk 1 above)
Comment (anton-lisanin) [reply proxy/common_neon/solana_interactor.py]: Replaced by the single request with the JSON RPC batch

File: proxy/common_neon/solana_interactor.py (hunk 2 above)
Comment (anton-lisanin) [reply proxy/common_neon/solana_interactor.py]: Done

File: proxy/common_neon/solana_interactor.py (hunk 5 above)
Comment (anton-lisanin) [reply proxy/common_neon/solana_interactor.py]: `confirm_transaction` removed

File: proxy/common_neon/solana_interactor.py [hunk 6]
```diff
@@ -91,8 +91,19 @@ def get_sol_balance(self, account):
...
+            params = (size, {"commitment": "confirmed"})
+            request_id = next(self.client._provider._request_counter) + 1
+
+            request = {"jsonrpc": "2.0", "id": request_id, "method": "getMinimumBalanceForRentExemption", "params": params}
+            request_data.append(request)
+
+        response = requests.post(self.client._provider.endpoint_uri, headers={"Content-Type": "application/json"}, json=request_data)
+        response.raise_for_status()
+
+        return list(map(lambda r: r["result"], response.json()))
```
Comment (s-medvedev) [inline proxy/common_neon/solana_interactor.py]: We should map results with requests by `request_id`. From https://www.jsonrpc.org/specification#batch:

> The Response objects being returned from a batch call MAY be returned in any order within the Array. The Client SHOULD match contexts between the set of Request objects and the resulting set of Response objects based on the id member within each Object.

File: proxy/common_neon/solana_interactor.py [hunk 7]
```diff
@@ -109,12 +150,29 @@ def send_transaction_unconfirmed(self, txn: Transaction):
...
+            params = (base64_transaction, {"skipPreflight": skip_preflight, "encoding": "base64", "preflightCommitment": "confirmed"})
+
+            request_id = next(self.client._provider._request_counter) + 1
+            request = {"jsonrpc": "2.0", "id": request_id, "method": "sendTransaction", "params": params}
+            request_data.append(request)
+
+        response = requests.post(self.client._provider.endpoint_uri, headers={"Content-Type": "application/json"}, json=request_data)
+        response.raise_for_status()
+
+        return list(map(lambda r: r["result"], response.json()))
```
Comment (s-medvedev) [inline proxy/common_neon/solana_interactor.py]: Map responses too.

File: proxy/common_neon/solana_interactor.py [hunk 8]
```diff
@@ -134,30 +192,57 @@ def get_measurements(self, result):
...
+
+        response = requests.post(self.client._provider.endpoint_uri, headers={"Content-Type": "application/json"}, json=request_data)
+        response.raise_for_status()
 
-    def collect_results(self, receipts, eth_trx=None, reason=None):
-        results = []
-        for rcpt in receipts:
-            results.append(self.collect_result(rcpt, eth_trx, reason))
-        return results
+        return cast(List[RPCResponse], response.json())
```
Comment (s-medvedev) [inline proxy/common_neon/solana_interactor.py]: Map responses too.

**Review by s-medvedev — COMMENTED** [review]
Looks good. Merge after getting openzeppelin results.

**Review by s-medvedev — APPROVED** [review]

**Pull request merged**